# Section 3: Classification Supervisée — Modélisation et Évaluation

Dans cette section, nous évaluons la performance de plusieurs algorithmes de classification supervisée sur le jeu de données *Women's E-Commerce Clothing Reviews*. 

L'objectif principal est de prédire la recommandation d'achat (`Recommended IND`) à partir des caractéristiques textuelles (TF-IDF) et numériques des avis clients.




### 3.1. Pipeline d'Entraînement et Stratégie de Validation
L'ensemble des modèles est évalué à l'aide d'une **Validation Croisée Stratifiée à 5 plis (`StratifiedKFold`, $k=5$)** et optimisé via une recherche sur grille (`GridSearchCV`). La métrique principale retenue pour l'optimisation est le **F1-Score (weighted)** afin de gérer le déséquilibre éventuel des classes.

In [1]:
import sys
sys.path.append('..')

import pandas as pd
import scipy.sparse as sp
from pathlib import Path
from src.train import run_model_pipeline


DATA_PATH = Path('../data/processed')

X_train = sp.load_npz(DATA_PATH / 'X_train_processed.npz')
X_test = sp.load_npz(DATA_PATH / 'X_test_processed.npz')
y_train = pd.read_csv(DATA_PATH / 'y_train.csv').values.ravel()
y_test = pd.read_csv(DATA_PATH / 'y_test.csv').values.ravel()


df_results = run_model_pipeline(X_train, y_train, X_test, y_test)
display(df_results)

Entraînement de : K-Nearest Neighbors...
Entraînement de : Multinomial Naïve Bayes...
Entraînement de : Decision Tree...
Entraînement de : Random Forest...
Entraînement de : SVM...


c:\Users\Administrateur\venv\Lib\site-packages\sklearn\svm\_base.py:236: FutureWarning: The `probability` parameter was deprecated in 1.9 and will be removed in version 1.11. Use `CalibratedClassifierCV(SVC(), ensemble=False)` instead of `SVC(probability=True)`
  warnings.warn(


,Modèle,Accuracy,F1-Score,ROC-AUC,Meilleurs Paramètres
0,SVM,0.939911,0.939882,0.976073,"{'C': 0.5, 'kernel': 'linear'}"
1,Random Forest,0.934157,0.932819,0.972103,"{'max_depth': None, 'n_estimators': 100}"
2,Decision Tree,0.929683,0.932725,0.966733,"{'criterion': 'gini', 'max_depth': 5}"
3,K-Nearest Neighbors,0.919455,0.915702,0.960037,"{'n_neighbors': 11, 'weights': 'uniform'}"
4,Multinomial Naïve Bayes,0.905178,0.904702,0.953054,{'alpha': 0.1}


### 3.2. Analyse et Interprétation des Résultats

Le tableau comparatif ci-dessus récapitule les métriques finales (**Accuracy**, **F1-Score**, et **ROC-AUC**) obtenues sur l'ensemble de test pour chaque modèle.

#### 1. Performance Globale
* **Support Vector Machine (SVM)** obtient les meilleurs résultats globaux avec une **Accuracy de 93.99%**, un **F1-Score de 0.9399** et un **ROC-AUC de 0.9761**. Le noyau linéaire (`kernel='linear'`) s'avère particulièrement efficace pour traiter la grande dimensionnalité des matrices TF-IDF.
* **Random Forest** et **Decision Tree** suivent de près avec des F1-Scores respectifs de **0.9328** et **0.9327**.
* **$k$-Nearest Neighbors ($k$-NN)** : Affiche une solide performance avec **91.95% d'Accuracy** et un **F1-Score de 0.9157**. La configuration optimale utilise $k=11$ voisins avec une pondération uniforme (`weights='uniform'`).
* **Multinomial Naïve Bayes** : Réalise une **Accuracy de 90.52%** et un **F1-Score de 0.9047** avec un lissage de Laplace ($\alpha=0.1$). Bien qu'il se situe légèrement en dessous des modèles complexes, Naïve Bayes reste le modèle le plus rapide à entraîner et offre une excellente ligne de base pour le traitement du langage naturel.